tailscale/resource_oauth_client.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) David Bond, Tailscale Inc, & Contributors
+// SPDX-License-Identifier: MIT
+
+package tailscale
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/go-cty/cty"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"tailscale.com/client/tailscale/v2"
+)
+
+func resourceOAuthClient() *schema.Resource {
+	return &schema.Resource{
+		Description:   "The oauth_client resource allows you to create OAuth clients to programmatically interact with the Tailscale API.",
+		ReadContext:   resourceOAuthClientRead,
+		CreateContext: resourceOAuthClientCreate,
+		DeleteContext: resourceOAuthClientDelete,
+		UpdateContext: schema.NoopContext,
+		// Importer: &schema.ResourceImporter{StateContext: schema.ImportStatePassthroughContext}, no import support - the key is not returned by the API so it'd serve no purpose
+		Schema: map[string]*schema.Schema{
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "A description of the key consisting of alphanumeric characters. Defaults to `\"\"`.",
+				ForceNew:    true,
+				ValidateDiagFunc: func(i interface{}, p cty.Path) diag.Diagnostics {
+					if len(i.(string)) > 50 {
+						return diagnosticsError(nil, "description must be 50 characters or less")
+					}
+					return nil
+				},
+			},
+			"scopes": {
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Required:    true,
+				Description: "Scopes to grant to the client. See https://tailscale.com/kb/1215/ for a list of available scopes.",
+				ForceNew:    true,
+			},
+			"tags": {
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional:    true,
+				Description: "A list of tags that access tokens generated for the OAuth client will be able to assign to devices. Mandatory is the scopes include \"devices:core\" or \"auth_keys\".",
+				ForceNew:    true,
+			},
+			"id": {
+				Type:        schema.TypeString,
+				Description: "The client ID, also known as the key id. Used with the client secret to generate access tokens.",
+				Computed:    true,
+				Sensitive:   true,
+			},
+			"key": {
+				Type:        schema.TypeString,
+				Description: "The client secret, also known as the key. Used with the client ID to generate access tokens.",
+				Computed:    true,
+				Sensitive:   true,
+			},
+			"created_at": {
+				Type:        schema.TypeString,
+				Description: "The creation timestamp of the key in RFC3339 format",
+				Computed:    true,
+			},
+			"user_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "ID of the user who created this key, or empty if not created by a user.",
```

Comment:
Think it might be helpful to add some colour around when this is the case:

```suggestion
				Description: "ID of the user who created this key, empty for OAuth clients created by other OAuth clients.",
```